cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.cpp

```diff
@@ -0,0 +1,413 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+inline std::int64_t max_wg(const sycl::queue& q) {
+    return q.get_device().template get_info<sycl::info::device::max_work_group_size>();
+}
+
+inline std::int64_t local_mem_size(const sycl::queue& q) {
+    return q.get_device().template get_info<sycl::info::device::local_mem_size>();
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+class kernel_reduction_rm_cw_inplace {
+    using inp_t = const Float*;
+    using out_t = Float*;
+
+public:
+    kernel_reduction_rm_cw_inplace(inp_t const input_,
+                                   out_t const output_,
+                                   const std::int64_t height_,
+                                   const std::int32_t lstride_,
+                                   const BinaryOp& binary_,
+                                   const UnaryOp& unary_)
+            : input{ input_ },
+              output{ output_ },
+              unary{ unary_ },
+              binary{ binary_ },
+              height{ height_ },
+              lstride{ lstride_ } {}
```

Comment:
Some checks on input parameters?

Reply:
It is a kernel. And there is no way it can be called outside